libp2p/src/builder.rs

```diff
@@ -0,0 +1,282 @@
+{
+    type E: Send + Sync + 'static;
+    type D: Send;
+    type U: Send;
+}
+
+impl<T> BuildableTransport for T
+where
+    T: Transport<Output = (libp2p_identity::PeerId, StreamMuxerBox)> + Send + Unpin + 'static,
+    <T as Transport>::Error: Send + Sync + 'static,
+    <T as Transport>::Dial: Send,
+    <T as Transport>::ListenerUpgrade: Send,
+{
+    type E = T::Error;
+    type D = T::Dial;
+    type U = T::ListenerUpgrade;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp"))]
+    fn tcp() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .without_relay()
+                .no_more_other_transports()
+                .without_dns()
+                .build();
+    }
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp", feature = "relay"))]
+    fn tcp_relay() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let (relay_transport, _) = libp2p_relay::client::new(key.public().to_peer_id());
+
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .with_relay(relay_transport)
+                .no_more_other_transports()
+                .without_dns()
+                .build();
+    }
+
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp", feature = "dns"))]
+    fn tcp_dns() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .without_relay()
+                .no_more_other_transports()
+                .with_dns()
+                .build();
+    }
+
+    /// Showcases how to provide custom transports unknown to the libp2p crate, e.g. QUIC or WebRTC.
+    #[test]
+    #[cfg(all(feature = "tokio", feature = "tcp"))]
+    fn tcp_other_transport_other_transport() {
+        let key = libp2p_identity::Keypair::generate_ed25519();
+        let _: libp2p_core::transport::Boxed<(libp2p_identity::PeerId, StreamMuxerBox)> =
+            TransportBuilder::new(key)
+                .with_tokio()
+                .with_tcp()
+                .without_relay()
+                .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
+                .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
+                .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
+                .no_more_other_transports()
+                .without_dns()
+                .build();
+    }
```

Comment:
`TransportBuilder` allows users to bring their own `Transport` implementations, e.g. `libp2p-quic`, even though they are unknown to `libp2p`. The only requirement is that they implement `Transport<Output = (PeerId, StreamMuxerBox)>`.

Reply:
This is awesome!! 

I has also discovered that I had to do some shuffling with Transports if I wanted more than 2 fallbacks (dns, quic, and/or webrtc), like this:

https://github.com/DougAnderson444/libp2peasy/blob/ede291d7fc1554549a954c05927d1eb5051802fc/src/transport.rs#L64-L74 

Does this builder mean we wouldn't need to do that anymore? (If so, great, because `Either, Either...` types were a bit hard to keep track of)

The other idea I had was to "pick some sane defaults" for typical use, then give the option to pass in a custom Config overriding it but you've likely already considered that in here somewhere.

Reply:
> The other idea I had was to "pick some sane defaults" for typical use, then give the option to pass in a custom Config overriding it but you've likely already considered that in here somewhere. 

Whilst sounding appealing, this is quite the footgun due to how cargo features work.

You shouldn't have features that change functionality because any additional dependency in your tree could suddenly change behaviour in your code.

Thus, the better idea is to just make the builder as easy as possible but have the user explicitly say what they want.

Reply:
> I has also discovered that I had to do some shuffling with Transports if I wanted more than 2 fallbacks (dns, quic, and/or webrtc), like this:
> 
> https://github.com/DougAnderson444/libp2peasy/blob/ede291d7fc1554549a954c05927d1eb5051802fc/src/transport.rs#L64-L74 

Yes that is a known issue and will be addressed by this solution.